WIP - hack the disassembler

dmpots opened the pull request.

clayborg reviewed (changes requested):
I checked in:
```
commit 078dd351d8eb7cd53f0bb810a6c57b2bc6ba6015 (HEAD -> llvm-server-plugins, origin/llvm-server-plugins)
Author: Greg Clayton <gclayton@fb.com>
Date:   Tue Aug 5 10:38:49 2025 -0700

    Add full ArchSpec support for AMD.
    
    Added full ArchSpec support for all GPU variants. This fixes 32 bit r600
    support and also creates a new ArchSpec::Core definitions for every AMD
    GPU. I modified the ObjectFileELF.cpp to decode the exact GPU core by
    parsing the information in the ELF header.
```
which fixes all of the AMD GPU stuff in ArchSpec. Can you rebase this on top of that and then ping me with what is left?

dmpots commented:
> I checked in:
> 
> ```
> commit 078dd351d8eb7cd53f0bb810a6c57b2bc6ba6015 (HEAD -> llvm-server-plugins, origin/llvm-server-plugins)
> Author: Greg Clayton <gclayton@fb.com>
> Date:   Tue Aug 5 10:38:49 2025 -0700
> 
>     Add full ArchSpec support for AMD.
>     
>     Added full ArchSpec support for all GPU variants. This fixes 32 bit r600
>     support and also creates a new ArchSpec::Core definitions for every AMD
>     GPU. I modified the ObjectFileELF.cpp to decode the exact GPU core by
>     parsing the information in the ELF header.
> ```
> 
> which fixes all of the AMD GPU stuff in ArchSpec. Can you rebase this on top of that and then ping me with what is left?

@clayborg I rebased on that change and updated the PR. Please take a look, thanks!

clayborg commented on lldb/source/Target/Target.cpp:
```diff
@@ -4557,7 +4557,17 @@ const char *TargetProperties::GetDisassemblyCPU() const {
   llvm::StringRef str = GetPropertyAtIndexAs<llvm::StringRef>(
       idx, g_target_properties[idx].default_cstr_value);
-  return str.empty() ? nullptr : str.data();
+
+  // If the user has an explict setting, then use that.
+  if (!str.empty())
+    return str.data();
+
+  // Next, check if the architecture has a default CPU setting.
+  std::string cpu = m_target->GetArchitecture().GetClangTargetCPU();
+  if (!cpu.empty())
+    return ConstString(cpu).GetCString();
+
+  return nullptr;
```
I wouldn't put this into this function as we want the `GetDisassemblyCPU()` call to always get the properties from the properties only without any additional work. I don't think any other properties accessors do extra work like this. Maybe we should move this blurb of code inside a new `lldb_private::Target` method:
```
llvm::StringRef Target::GetDisassemblyCPUName() const {
  // Try to get the GPU name from the target settings first.
  llvm::StringRef cpu(GetDisassemblyCPU());
  if (!cpu.empty())
    return cpu;
  // Next, check if the architecture has a default CPU setting.
  std::string cpu = m_target->GetArchitecture().GetClangTargetCPU();
  if (!cpu.empty())
    return ConstString(cpu).GetCString();

  return nullptr;
}
```

dmpots replied:
@clayborg  I reverted the change to `GetDisassemblyCPU()` and modified the Target code to update the property when the target architecture is changed: [ede48b1](https://github.com/clayborg/llvm-project/pull/24/commits/ede48b1df0341b96c5155a4f03d44a527e3baf28).

Let me know what you think.

dmpots commented:
Most of the target properties do not do extra work. I found one example of GetEnvironment() that calls ComputeEnviornment(): https://github.com/clayborg/llvm-project/blob/b8195e3a8e77ec15f6abbcce86d6a51dca13a5c4/lldb/source/Target/Target.cpp#L4607

So I think it makes sense to move the code out of the TargetProperties getter, but then we end up with two methods on the Target class that have very similar names (Target inherit from TargetProperties).

There are a number of calls to GetDisassemblyCPU that directly hit the target
```
lldb/source/API/SBFunction.cpp:          target_sp->GetDisassemblyCPU(), target_sp->GetDisassemblyFeatures(),
lldb/source/API/SBSymbol.cpp:            target_sp->GetDisassemblyCPU(), target_sp->GetDisassemblyFeatures(),
lldb/source/API/SBTarget.cpp:          target_sp->GetDisassemblyCPU(), target_sp->GetDisassemblyFeatures(),
lldb/source/API/SBTarget.cpp:          target_sp->GetDisassemblyCPU(), target_sp->GetDisassemblyFeatures(),
lldb/source/API/SBTarget.cpp:        target_sp->GetDisassemblyCPU(), target_sp->GetDisassemblyFeatures(),
lldb/source/Commands/CommandObjectDisassemble.cpp:    if (const char *cpu = target->GetDisassemblyCPU())
lldb/source/Core/Disassembler.cpp:    cpu = target.GetDisassemblyCPU();
lldb/source/Core/DumpDataExtractor.cpp:        target_sp->GetDisassemblyCPU(), target_sp->GetDisassemblyFeatures(),
lldb/source/Target/StackFrame.cpp:                target->GetDisassemblyCPU(), target->GetDisassemblyFeatures(),
```

So if we have both GetDisassemblyCPU and GetDisassemblyCPUName that seems rather confusing. We could rename the existing GetDisassemblyCPU to GetDisassemblyCPUTargetProperty and then add a new GetDisassemblyCPU with the new logic.

What do you think?
